File: go/1461.go
package main

import (
	"math"
	"strconv"
)

func hasAllCodes(s string, k int) bool {
	table := map[int]int{}

	for idx := 0; idx <= len(s)-k; idx++ {
		sub := s[idx : idx+k]
		val, _ := strconv.ParseInt(sub, 2, 32)

		ct, ok := table[int(val)]
		if ok {
			table[int(val)] = ct + 1
		} else {
			table[int(val)] = 1
		}
	}

	exp := math.Pow(2, float64(k))

	for val := 0; val < int(exp); val++ {
		_, ok := table[val]
		if !ok {
			return false
		}
	}

	return true
}

File: zig/1461.zig
const std = @import("std");

fn solution(s: []const u8, k: usize, alloc: std.mem.Allocator) !bool {
    var table: std.StringHashMap(bool) = .init(alloc);
    defer table.deinit();

    for (0..s.len - k + 1) |idx| {
        const sub = s[idx .. idx + k];
        try table.put(sub, true);
    }

    return table.count() == std.math.pow(u64, 2, k);
}

const expect = std.testing.expect;

test "test case #1" {
    const s = "00110110";
    const k = 2;

    const alloc = std.testing.allocator;

    const expected = true;
    const sol = try solution(s, k, alloc);

    try expect(sol == expected);
}

test "test case #2" {
    const s = "0110";
    const k = 1;

    const alloc = std.testing.allocator;

    const expected = true;
    const sol = try solution(s, k, alloc);

    try expect(sol == expected);
}

test "test case #3" {
    const s = "0110";
    const k = 2;

    const alloc = std.testing.allocator;

    const expected = false;
    const sol = try solution(s, k, alloc);

    try expect(sol == expected);
}
